File: extensions/vscode/src/extension/services/__tests__/configParse.test.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 alibaba/open-code-review Contributors

// src/extension/services/__tests__/configParse.test.ts
import { parseConfig, toConfigSetArgs } from '../configParse';

describe('parseConfig', () => {
  it('converts a complete config to camelCase', () => {
    const raw = JSON.stringify({
      provider: 'anthropic',
      providers: {
        anthropic: { api_key: 'k', model: 'claude-opus-4-6', models: ['claude-opus-4-6'] },
      },
      custom_providers: {
        'my-llm': { url: 'https://x', protocol: 'openai', model: 'm', api_key: 'k2' },
      },
      llm: { url: 'u', auth_token: 't', model: 'm', use_anthropic: true, auth_header: 'x-api-key' },
      language: 'Chinese',
    });
    expect(parseConfig(raw)).toEqual({
      provider: 'anthropic',
      model: '',
      providers: {
        anthropic: { apiKey: 'k', url: '', protocol: '', model: 'claude-opus-4-6', models: ['claude-opus-4-6'], authHeader: '' },
      },
      customProviders: {
        'my-llm': { apiKey: 'k2', url: 'https://x', protocol: 'openai', model: 'm', authHeader: '' },
      },
      llm: { url: 'u', authToken: 't', model: 'm', useAnthropic: true, authHeader: 'x-api-key' },
      language: 'Chinese',
    });
  });

  it('provides defaults for missing fields', () => {
    const cfg = parseConfig('{}');
    expect(cfg?.llm.url).toBe('');
    expect(cfg?.llm.useAnthropic).toBe(true);
    expect(cfg?.providers).toEqual({});
    expect(cfg?.customProviders).toEqual({});
    expect(cfg?.language).toBe('Chinese');
  });

  it('returns null for an empty string', () => {
    expect(parseConfig('')).toBeNull();
  });

  it('ignores providers when it is an array', () => {
    const cfg = parseConfig(JSON.stringify({ providers: ['bad'], custom_providers: [] }));
    expect(cfg?.providers).toEqual({});
    expect(cfg?.customProviders).toEqual({});
  });
});

describe('toConfigSetArgs', () => {
  it('builds config set arguments', () => {
    expect(toConfigSetArgs('llm.model', 'opus')).toEqual(['config', 'set', 'llm.model', 'opus']);
    expect(toConfigSetArgs('providers.anthropic.api_key', 'sk')).toEqual(['config', 'set', 'providers.anthropic.api_key', 'sk']);
  });
});

File: extensions/idea/src/test/kotlin/com/alibaba/opencodereview/idea/services/ConfigParseTest.kt
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 alibaba/open-code-review Contributors

package com.alibaba.opencodereview.idea.services

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertNull
import kotlin.test.assertTrue

class ConfigParseTest {

    @Test
    fun `empty content returns null`() {
        assertNull(parseConfig(""))
        assertNull(parseConfig("   \n  "))
    }

    @Test
    fun `snake_case configuration becomes a camelCase domain model`() {
        val cfg = parseConfig(
            """
            {
              "provider": "anthropic",
              "model": "claude-opus-4-8",
              "language": "English",
              "providers": {
                "anthropic": {
                  "api_key": "sk-x",
                  "url": "https://api.anthropic.com",
                  "protocol": "anthropic",
                  "model": "claude-opus-4-8",
                  "auth_header": "x-api-key",
                  "models": ["a", "b"]
                }
              },
              "custom_providers": {
                "mine": { "api_key": "k", "url": "http://localhost" }
              },
              "llm": {
                "url": "http://llm",
                "auth_token": "t",
                "model": "m",
                "auth_header": "h"
              }
            }
            """.trimIndent(),
        )!!

        assertEquals("anthropic", cfg.provider)
        assertEquals("English", cfg.language)

        val preset = cfg.providers.getValue("anthropic")
        // Guard against incorrect field names: snake_case names would silently leave the frontend blank.
        assertEquals("sk-x", preset.apiKey)
        assertEquals("x-api-key", preset.authHeader)
        assertEquals(listOf("a", "b"), preset.models)

        val custom = cfg.customProviders.getValue("mine")
        assertEquals("k", custom.apiKey)
        assertEquals("", custom.protocol)

        assertEquals("http://llm", cfg.llm.url)
        assertEquals("t", cfg.llm.authToken)
        assertEquals("h", cfg.llm.authHeader)
    }

    @Test
    fun `missing fields fall back to empty strings without throwing`() {
        val cfg = parseConfig("{}")!!
        assertEquals("", cfg.provider)
        assertEquals("", cfg.model)
        assertEquals(emptyMap(), cfg.providers)
        assertEquals(emptyMap(), cfg.customProviders)
        assertEquals("", cfg.llm.url)
    }

    @Test
    fun `missing language defaults to Chinese`() {
        assertEquals("Chinese", parseConfig("{}")!!.language)
        assertEquals("Chinese", parseConfig("""{"language": ""}""")!!.language)
        assertEquals("English", parseConfig("""{"language": "English"}""")!!.language)
    }

    @Test
    fun `use_anthropic is disabled only by an explicit boolean false`() {
        assertTrue(parseConfig("{}")!!.llm.useAnthropic)
        assertTrue(parseConfig("""{"llm": {}}""")!!.llm.useAnthropic)
        assertTrue(parseConfig("""{"llm": {"use_anthropic": true}}""")!!.llm.useAnthropic)
        assertEquals(false, parseConfig("""{"llm": {"use_anthropic": false}}""")!!.llm.useAnthropic)
    }

    @Test
    fun `string false does not disable use_anthropic`() {
        // In JavaScript, "false" !== false is true, so the option stays enabled.
        // Using booleanOrNull here would incorrectly disable it and diverge from frontend behavior.
        assertTrue(parseConfig("""{"llm": {"use_anthropic": "false"}}""")!!.llm.useAnthropic)
    }

    @Test
    fun `missing models is null rather than an empty list`() {
        // Null means unconfigured, so the frontend falls back to preset models.
        // An empty list means configured with no entries and renders an empty dropdown.
        val entry = parseConfig("""{"providers": {"anthropic": {}}}""")!!.providers.getValue("anthropic")
        assertNull(entry.models)
    }

    @Test
    fun `non-string elements in models are filtered out`() {
        val entry = parseConfig("""{"providers": {"p": {"models": ["a", 1, null, "b"]}}}""")!!
            .providers.getValue("p")
        assertEquals(listOf("a", "b"), entry.models)
    }

    @Test
    fun `incorrect field types are treated as missing`() {
        val cfg = parseConfig("""{"provider": 123, "model": true, "providers": []}""")!!
        assertEquals("", cfg.provider)
        assertEquals("", cfg.model)
        assertEquals(emptyMap(), cfg.providers)
    }

    @Test
    fun `unknown top-level fields do not affect parsing`() {
        val cfg = parseConfig("""{"provider": "openai", "future_flag": {"deep": [1]}}""")!!
        assertEquals("openai", cfg.provider)
    }

    @Test
    fun `invalid JSON propagates an exception`() {
        // ConfigService.read() catches this as unconfigured; the parser itself does not swallow the exception.
        assertFailsWith<Exception> { parseConfig("{ not json") }
    }

    @Test
    fun `config set argument order is fixed`() {
        assertEquals(listOf("config", "set", "provider", "openai"), toConfigSetArgs("provider", "openai"))
    }
}
